internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go

```diff
@@ -0,0 +1,241 @@
+}
+
+// dynamicScaleDownExecutor is able to perform dynamic ScaleDown execution actions
+type dynamicScaleDownExecutor interface {
+	ScaleDown() error
+}
+
+// dynamicScaleDownEvaluator is able to perform dynamic ScaleDown evaluation actions
+type dynamicScaleDownEvaluator interface {
+	ShouldScaleDown() (bool, error)
+}
+
+// dynamicScaleDownProcessor is able to process dynamic ScaleDown reconcile events
+type dynamicScaleDownProcessor interface {
+	dynamicScaleDownExecutor
+	dynamicScaleDownEvaluator
+}
+
+// noopDynamicScaleDownProcessor is a dynamicScaleDownProcessor where
+// the scale down is a noop and it always returns that no scale down action
+// is needed
+type noopDynamicScaleDownProcessor struct {
+}
+
+var _ dynamicScaleDownEvaluator = &noopDynamicScaleDownProcessor{}
+var _ dynamicScaleDownExecutor = &noopDynamicScaleDownProcessor{}
+var _ dynamicScaleDownProcessor = &noopDynamicScaleDownProcessor{}
+
+func (p *noopDynamicScaleDownProcessor) ScaleDown() error {
+	return nil
+}
+
+func (p *noopDynamicScaleDownProcessor) ShouldScaleDown() (bool, error) {
+	return false, nil
+}
+
+// standardDynamicScaleDownProcessor is the default dynamicScaleDownProcessor
+// used when dynamic scaling is enabled.
+// It assumes the provided kafkaStreamingUnitCountPerClusterList does not
+// contain any element with a Status attribute with value 'failed'
+type standardDynamicScaleDownProcessor struct {
+	clusterID string
+	// kafkaStreamingUnitCountPerClusterList must not contain any element
+	// with a Status attribute with value 'failed'
+	kafkaStreamingUnitCountPerClusterList services.KafkaStreamingUnitCountPerClusterList
+	supportedKafkaInstanceTypesConfig     *config.SupportedKafkaInstanceTypesConfig
+	clusterService                        services.ClusterService
+
+	// dryRun controls whether the ScaleDown method performs real actions.
+	// Useful when you don't want to trigger a real scale down.
+	dryRun bool
+}
+
+var _ dynamicScaleDownEvaluator = &standardDynamicScaleDownProcessor{}
+var _ dynamicScaleDownExecutor = &standardDynamicScaleDownProcessor{}
+var _ dynamicScaleDownProcessor = &standardDynamicScaleDownProcessor{}
+
+// ShouldScaleDown indicates whether a new data plane
+// cluster should be created for a given instance type in the given provider
+// and region.
+// It returns true if all the following conditions happen:
+// 1. If specified the cluster is empty i.e it does not contain any streaming unit
+// 2. If the cluster can be removed without triggering a scale down action
+// Otherwise false is returned.
+// Note: This method assumes kafkaStreamingUnitCountPerClusterList does not
+//       contain elements with the Status attribute with the 'failed' value.
+//       Thus, if the type is constructed with the assumptions being true, it
+//       can be considered as the 'failed' state Clusters are not included in
+//       the calculations.
+func (p *standardDynamicScaleDownProcessor) ShouldScaleDown() (bool, error) {
```

Comment:
As mentioned in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1266#discussion_r957451954 let's think about the sibling clusters logic: with the new dynamic scaling behavior is that logic still needed at all or are we already dealing with it somehow? maybe there's a better way to ensure we don't end up with 0 cluters in the region that does not require changing the state of already existing clusters to ready? like, not setting the cluster to deprovision at all if we detect that would end up happening. Alternatively, if we end up determining that we still need that same sibling logic, where should be more appropriate to put that?

Reply:
> like, not setting the cluster to deprovision at all if we detect that would end up happening

This is already taken care of by https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1267/files#diff-7750701a0ae3440f5aacd7bdbe12d43b25c1c2f01eaa827dd85f10a10aa39c95R130

i.e after the evaluation of all the scaling down logic has been completed. We are not going to mark the scale for deprovisioning if either of the checks returned false i.e 
if one of the following conditions are true, then the cluster shouldn't be deleted: 
- the cluster is non empty 
- or we do not have a sibling cluster in the region after the cluster is deleted, 
- we'll be eating the slack capacity after the cluster is deleted 
- or we'll be unable to provision the largest streaming unit in the region after the cluster is deleted.  

> maybe there's a better way to ensure we don't end up with 0 cluters in the region that does not require changing the state back of already existing clusters to ready?

Because of the checks we've, I am not so much worried about the sibling clusters cases and we can remove the logic we discussed in https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1266#discussion_r957451954

The case where I was thinking that bringing back the state of the cluster will be useful is when:
- we saw that we can delete a cluster, 
- but before we mark it as empty, a Kafka is assigned onto that cluster: an event we do not know about so we end up marking the cluster for deletion anyway.  

However, thinking afresh about this, an SQL query might be enough when marking the cluster for deletion. 
```
set status=deprovisioning where given_cluster_id not in (select cluster_id from kafka_requests).
```
I'll give it more thought today but I feel like that should be enough and we can delete the other logic effectively.

Reply:
discussed internally:

The decision was to keep the check that the cluster is empty, if not empty then bring back the state of the cluster to `ready`, otherwise we go ahead and delete the cluster from OCM. 
I've updated the code in that direction.